modules/openid-federation-common/src/commonMain/kotlin/com/sphereon/oid/fed/common/mime/JsonUrlEncoder.kt

```diff
@@ -0,0 +1,131 @@
+package com.sphereon.oid.fed.common.mime
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonElement
+
+private val qpAllowedChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.~".toSet()
+
+/**
+ * URL encode a String.
+ * Converts characters not allowed in URL query parameters to their percent-encoded equivalents.
+ *
+ * @return URL encoded String
+ */
+fun urlEncodeValue(input: String): String {
+    return buildString {
+        input.forEach { char ->
+            if (char in qpAllowedChars) {
+                append(char)
+            } else {
+                append('%')
+                append(char.code.toString(16).uppercase().padStart(2, '0'))
+            }
+        }
+    }
+}
+
+/**
+ * Extension function to URL encode a String.
+ * Converts characters not allowed in URL query parameters to their percent-encoded equivalents.
+ *
+ * @return URL encoded String
+ */
+fun String.toUrlEncodedValue(): String {
+    return urlEncodeValue(this)
+}
+
+/**
+ * Extension function to URL encode a JsonElement.
+ * Converts the JsonElement to a JSON String and then URL encodes it.
+ *
+ * @return URL encoded String representation of the JsonElement
+ */
+fun JsonElement.toUrlEncodedValue(): String {
+    return this.toString().toUrlEncodedValue()
+}
+
+/**
+ * Inline function to URL encode any serializable object.
+ * Converts the object to a JSON string and then URL encodes it.
+ *
+ * @return URL encoded JSON String representation of the object
+ */
+inline fun <reified T> T.toUrlEncodedJsonValue(): String {
+    return Json.encodeToString(this).toUrlEncodedValue()
+}
+
+/**
+ * Function to URL encode any serializable object using a provided serializer.
+ * Converts the object to a JSON string using the serializer and then URL encodes it.
+ *
+ * @param serializer The serializer to use for converting the object to a JSON string
+ * @return URL encoded JSON String representation of the object
+ */
+fun <T> T.toUrlEncodedJsonValue(serializer: KSerializer<T>): String {
+    return Json.encodeToString(serializer, this).toUrlEncodedValue()
+}
+
+/**
+ * Extension function to decode a URL encoded String.
+ * Converts percent-encoded characters back to their original form.
+ *
+ * @return Decoded String
+ */
+fun String.fromUrlEncodedValue(): String {
```

Comment:
Please create a regular function with @JSexport first, then call that from this extension function. This allows the most flexibility in deciding whether we want to make the extension functions available to javascript for instance, or whether to only expose the regular functions

Reply:
Fixed in next commit